grails-app/services/org/pih/warehouse/data/PersonService.groovy

```diff
@@ -11,33 +11,136 @@ package org.pih.warehouse.data
         String[] names = extractNames(combinedNames)
-        return getPersonByNames(names)
+        List<Person> people = Person.findAllByFirstNameAndLastName(names[0], names[1])
+        if (people?.empty) {
+            return null
+        }
+
+        // We give preference to active people, so if there is one found, return it, otherwise return anyone.
+        return people.find{ it.active } ?: people.first()
+    }
+
+    /**
+     * Fetch an active user via their full name. We require the combined name to be provided (instead of first name
+     * and last name individually) because while internally we split the name so that it conforms to the firstName and
+     * lastName fields in the Person table, we want to eventually move away from forcing that name structure upon users.
+     *
+     * @param combinedNames The person's full name (first name + last name) combined into a single String.
+     * @return A Person who is active and has a firstName and lastName matching the combinedNames
+     */
+    Person getActivePersonByName(String combinedNames) {
+        if (StringUtils.isBlank(combinedNames)) {
+            return null
+        }
+
+        String[] names = extractNames(combinedNames)
+        return Person.findByActiveAndFirstNameAndLastName(true, names[0], names[1])
+    }
+
+    /**
+     * @param recipient Either a RFC822O internet/email address (ex: "Justin Miranda <justin@openboxes.com>")
+     *        or simply the recipient's name (ex: "Justin Miranda"). We accept both formats since either
+     *        format can be provided during data imports.
+     * @return A Person who has the provided name and email. Will be a new Person if they didn't previously exist.
+     */
+    Person getOrCreatePersonByRecipient(String recipient) {
+        if (StringUtils.isBlank(recipient)) {
+            return null
+        }
+
+        InternetAddress internetAddress
+        try
+        {
+            internetAddress = new InternetAddress(recipient, false)
+        }
+        catch (AddressException ignored) {
+            // If recipient isn't a valid internet address, it must be a regular name.
+            return getOrCreatePersonFromNames(recipient)
+        }
+
+        // If a person exists with the given email, return them. Note that this can return inactive people so
+        // the caller must properly handle that case. We do this to avoid creating duplicate users.
+        Person person = getPersonByEmail(internetAddress.address)
+        if (person) {
+            return person
+        }
+
+        // Otherwise create a new person.
+        if (!internetAddress.personal) {
+            throw new RuntimeException("Cannot save new recipient without a name: ${recipient}")
+        }
+        String[] names = extractNames(internetAddress.personal)
+        person = new Person(firstName: names[0], lastName: names[1], email: internetAddress.address)
+        if (!person.save(flush: true)) {
```

Comment:
is this flush needed?

Reply:
probably not but it was in the original code so I didn't want to change it without knowing why it was added

Reply:
Theoretically it shouldn't be necessary. But practically, it depends on where this code is being invoked. For adding a single person to the system, probably not. But in the previous version of Grails, things were less predictable when we were persisting lots of data all at once. Like say in a data import. It might be possible to remove the flush now (and I would be happy with that refactoring) but I appreciate Evan's instinct to leave well enough alone for now. And if/when we do make the change we need to ensure we have some integration tests around it to make sure it works for all use cases:
* add a single person in an admin page 
* add a single person per line item and assign them as recipient on the shipment item

Again it should work without the flush, but we should perform a discovery before making changes. This might be a good scenario to add to the Cascading Behavior on Associations tech debt that we've discussed in the past.